contracts/Staking.sol

```diff
@@ -0,0 +1,487 @@
+        _stakeFor(_account, _amount, _data);
+    }
+
+    /**
+     * @notice Unstakes `_amount` tokens, returning them to the user
+     * @param _amount Number of tokens staked
+     * @param _data Used in Unstaked event, to add signalling information in more complex staking applications
+     */
+    function unstake(uint256 _amount, bytes _data) external {
+        // unstake 0 tokens makes no sense
+        require(_amount > 0, ERROR_AMOUNT_ZERO);
+
+        // transfer tokens
+        require(stakingToken.transfer(msg.sender, _amount), ERROR_TOKEN_TRANSFER);
+
+        // process Stake
+        _modifyStakeBalance(msg.sender, _amount, false);
+
+        // Update global history
+        _updateTotalStaked();
+
+        emit Unstaked(msg.sender, _amount, totalStakedFor(msg.sender), _data);
+    }
+
+    /**
+     * @notice Get the total amount of tokens staked by all users
+     * @return The total amount of tokens staked by all users
+     */
+    function totalStaked() external view returns (uint256) {
+        return totalStakedAt(getBlockNumber64());
+    }
+
+    /**
+     * @notice Get the token used by the contract for staking and locking
+     * @return The token used by the contract for staking and locking
+     */
+    function token() external view returns (address) {
+        return address(stakingToken);
+    }
+
+    /**
+     * @notice Check whether it supports history of stakes
+     * @return Always true
+     */
+    function supportsHistory() external pure returns (bool) {
+        return true;
+    }
+
+    /**
+     * @notice Get last time `_account` modified its staked balance
+     * @param _account Account requesting for
+     * @return Last block number when account's balance was modified
+     */
+    function lastStakedFor(address _account) external view returns (uint256) {
+        return stakedHistory[_account].lastUpdated();
+    }
+
+    /**
+     * @notice Lock `_amount` staked tokens and assign `_manager` as manager with `_data` as data, so they can not be unstaked
+     * @param _amount The amount of tokens to be locked
+     * @param _manager The manager entity for this particular lock. This entity will have full control over the lock, in particular will be able to unlock it
+     * @param _data Data to parametrize logic for the lock to be enforced by the manager
+     * @return The id of the newly created lock
+     */
+    function lock(uint256 _amount, address _manager, bytes _data) external returns (uint256 _lockId) {
+        // lock 0 tokens makes no sense
+        require(_amount > 0, ERROR_AMOUNT_ZERO);
+
+        // check enough unlocked tokens are available
+        require(_amount <= unlockedBalanceOf(msg.sender), ERROR_NOT_ENOUGH_BALANCE);
+
+        accounts[msg.sender].lastLockId++;
+        _lockId = accounts[msg.sender].lastLockId;
+        accounts[msg.sender].activeLockIds.push(_lockId);
+        Lock storage _lock = accounts[msg.sender].locks[_lockId];
+        _lock.amount = _amount;
```

Comment:
Yes, but I seem to recall @sohkai advising against it because of higher gas consumption. @sohkai can you confirm? (Maybe I'm confused with another case)

Reply:
I found the comment:
https://github.com/aragon/aragon-apps/pull/479/files#r231335088